### PR TITLE
fix(cli): start fixing files add

### PR DESCRIPTION
**Pull request by dignifiedquire**
Fixes https://github.com/ipfs/js-ipfs/issues/480

### CODE REVIEW

File: src/cli/commands/files/add.js [hunk 1]
```diff
@@ -55,44 +55,71 @@ module.exports = {
...
 
     utils.getIPFS((err, ipfs) => {
-      if (err) throw err
+      if (err) {
+        throw err
+      }
+
+      // pull-stream -> stream -> pull-stream
+      // is not a good idea, so if we have a pull interface
+      // available, use it
```
Comment (diasdavid) [inline src/cli/commands/files/add.js]: `s/comment/TODO revisit when interface-ipfs-core exposes a pull-streams API`

File: src/cli/commands/files/add.js [hunk 2]
```diff
@@ -55,44 +55,71 @@ module.exports = {
...
+        }
+      }
+
+      createAddStream((err, addStream) => {
+        if (err) {
+          throw err
+        }
+
+        glob(path.join(inPath, '/**/*'), (err, list) => {
+          if (err) throw err
```
Comment (diasdavid) [inline src/cli/commands/files/add.js]: curly

File: src/cli/commands/files/add.js [hunk 3]
```diff
@@ -55,44 +55,71 @@ module.exports = {
...
+          if (list.length === 0) list = [inPath]
```
Comment (diasdavid) [inline src/cli/commands/files/add.js]: curly

File: src/cli/commands/files/add.js [hunk 4]
```diff
@@ -55,44 +55,71 @@ module.exports = {
...
+      path: file.path.substring(index, file.path.length),
+      content: fs.createReadStream(file.path)
+    })),
+    addStream,
+    pull.map((file) => ({
+      hash: file.hash,
+      path: file.path
+    })),
+    pull.collect((err, added) => {
+      if (err) throw err
```
Comment (diasdavid) [inline src/cli/commands/files/add.js]: curly

File: src/cli/commands/files/add.js [hunk 5]
```diff
@@ -55,44 +55,71 @@ module.exports = {
...
+
+function addPipeline (index, addStream, list) {
+  pull(
+    zip(
+      pull.values(list),
+      pull(
+        pull.values(list),
+        paramap(fs.stat.bind(fs), 50)
+      )
+    ),
```
Comment (diasdavid) [inline src/cli/commands/files/add.js]: We `zip` this?

**Comment by diasdavid** [comment]
@dignifiedquire did you had the chance to try this add against a go-ipfs running daemon? (same thing for cat, and get)

File: src/cli/commands/files/add.js (hunk 3 above)
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: Please make a PR to aegir to enforce this if you are going to comment on every single curly

**Comment by dignifiedquire** [comment]
@diasdavid that's how I tested interop yesternight

File: src/cli/commands/files/add.js (hunk 3 above)
Comment (diasdavid) [inline src/cli/commands/files/add.js]: 👍

File: src/cli/commands/files/add.js (hunk 5 above)
Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: So I get a combo of the name + the stat below

Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: also zip is combining two streams, not zip as in compressing

Comment (diasdavid) [inline src/cli/commands/files/add.js]: oh, wanna make a comment on that line, it was clearly not obvious (I would expect that to be called concat)

Comment (dignifiedquire) [inline src/cli/commands/files/add.js]: concat is something entirely different `zip` is a standard name in function programming: https://en.wikipedia.org/wiki/Convolution_(computer_science)

**Comment by dignifiedquire** [comment]
@diasdavid should be ready now

**Pull request merged**